org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java

```diff
@@ -64,34 +77,52 @@ protected final void runToolingApiJob(IProgressMonitor monitor) {
-        request.standardOutput(processStreams.getOutput());
-        request.standardError(processStreams.getError());
-        request.standardInput(processStreams.getInput());
-        request.progressListeners(listeners.toArray(new ProgressListener[listeners.size()]));
-        request.cancellationToken(getToken());
-
-        // print the applied run configuration settings at the beginning of the console output
-        OutputStreamWriter writer = new OutputStreamWriter(processStreams.getConfiguration());
-        writeFixedRequestAttributes(fixedAttributes, writer, monitor);
-
-        // notify the listeners before executing the build launch request
-        Event event = new DefaultExecuteLaunchRequestEvent(processDescription, request);
-        CorePlugin.listenerRegistry().dispatch(event);
-
-        // launch the build
-        request.executeAndWait();
+        final FixedRequestAttributes attributes = getConfigurationAttributes().toFixedRequestAttributes();
+        ProjectConnection connection = null;
+        try {
+            // apply FixedRequestAttributes on connector and establish TAPI connection
+            GradleConnector connector = GradleConnector.newConnector().forProjectDirectory(attributes.getProjectDir());
+            GradleDistributionWrapper.from(attributes.getGradleDistribution()).apply(connector);
+            connector.useGradleUserHomeDir(attributes.getGradleUserHome());
+            connection = connector.connect();
+
+            // apply FixedRequestAttributes on build launcher
+            Launcher launcher = createLauncher(connection);
+            LongRunningOperation operation = launcher.getOperation();
+            operation.setJavaHome(attributes.getJavaHome());
+            operation.withArguments(attributes.getArguments());
+            operation.setJvmArguments(attributes.getJvmArguments());
+
+            // transient attributes
+            operation.setStandardOutput(processStreams.getOutput());
+            operation.setStandardError(processStreams.getError());
+            operation.setStandardInput(processStreams.getInput());
+            for (ProgressListener listener : listeners) {
+                operation.addProgressListener(listener);
+            }
+            operation.withCancellationToken(getToken());
+
+            // notify the listeners before executing the build launch request
+            Event event = new DefaultExecuteLaunchRequestEvent(processDescription, operation);
+            CorePlugin.listenerRegistry().dispatch(event);
+
+            // print the applied run configuration settings at the beginning of the console output
+            OutputStreamWriter writer = new OutputStreamWriter(processStreams.getConfiguration());
+            writeFixedRequestAttributes(attributes, writer, monitor);
+
+            // execute the build
+            launcher.run();
+        } finally {
+            if (connection != null) {
+                connection.close();
```

Comment:
Using a callback is a bit limiting. I see you had to copy-paste some code as a result instead of having it in the base class like before. It also forced you to create a new class (`GradleInvocation`), which is against our goal for removing indirection layers over the TAPI types.

Instead what I was thinking was returning a `BuildLauncher` or `TestLauncher` which automatically closes the underlying `ProjectConnection` after it has first been used. This can for instance be done with a listener that listens for build completion.